### PR TITLE
Changes to state manager and redis datastore schema

### LINKED ISSUE #951: Final state_manager update from worker happens too late
The state_manager is already removed (probably because of serialization) so the worker can't update Datastore/Redis after the Task has been completed.  This shows up in the logs as:

`No state_manager initialized, not updating Task info`

We should move this update to happen earlier.

### LINKED ISSUE #1030: Store Request and Group IDs in backend key value store
We currently use key/value stores as backend. This stores the key as 'TurbiniaTask:[task_id]' with a JSON blob as value containing the details (including request id and group id). If you want to query on request or group ID we loop through all keys and parse the value to get the relevant task IDs.

We could store the RequestID with a value list of TaskIDs and the GroupID with a value list of Request IDs to make status fetching faster.

### LINKED ISSUE #1239: Clean up Task statuses
Status is being set on the worker side to either `queued` or `running` in the run_wrapper block. Then the server side uses the celery stub to check the Task status via it's Celery States (STARTED, FAILURE, SUCCESS) although `STARTED` will never evaluate since `task_track_started` needs to be enabled context [here](https://docs.celeryq.dev/en/stable/userguide/configuration.html#task-track-started) then the API server checks for running in status or the successful flag to grab the current Task status. The `queued` state should also probably be removed or set differently/by the Server as a Task should never be in the `queued` state if we configured the celery worker to take on 1 Task per Worker. 

When we start to phase out PSQ, it'd be good to standardize how we're updating the Task status and how the Server and API Server check for the Task status. For example using celery states (which is available via `task.stub`) we can gather:
* PENDING: Task is waiting for execution or is in an unknown state.
* STARTED: Task has been started/executed on a Worker.
* SUCCESS: Task has been successfully executed.
* FAILURE: Task execution resulted in failure.

There is also `RETRY` or `REVOKED` and custom states that can be set/looked into, more details around celery states here: https://docs.celeryq.dev/en/stable/userguide/tasks.html#states

### LINKED ISSUE #1243: Improve Task progress tracking in Celery
With our move to using just Celery, Celery gives us some more functionality around how we give Task updates while the Worker is executing. 

References:
* https://docs.celeryq.dev/en/stable/userguide/calling.html#on-message
* https://docs.celeryq.dev/en/stable/userguide/tasks.html#custom-states

Using custom states/ the `update_state` method, come up with a better way to track a progress of a running Task. This may have to be different for each Task/Job depending what they are checking for and how long they are running, but some ideas are:
* Can have the progress metric be based on output file size for larger tasks such as Plaso
* Can have a default progress metric for Tasks that should be quick in nature (checking existence of a file)
* Can take the external programs latest stdout and provide it in the status update.
* Can take the latest modified timestamp of the output file

### LINKED ISSUE #1470: [Bug]: turbinia-client prints extraneous help output
### What steps will reproduce the bug?

Use `turbinia-client submit googleclouddisk --project <foo> --zone <bar> --disk_name <baz>` to submit processing request.

### What is the expected behavior?

Only the info about the correctly submitted processing request should be printed.

### What do you see instead?

The help output is displayed in the output even though the processing request gets submitted normally and all the required flags have been passed in.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by jleaniz**
### Description of the change

This pull request makes changes to the schema used to store Turbinia metadata in Redis. We are introducing a new key for Turbinia requests and changing TurbiniaTask keys to redis hashes instead of strings for optimization purposes.

Minor changes to the web UI.

Note: This is a breaking change. Redis databases from previous versions will not be compatible.

### Applicable issues

- fixes #1243
- fixes #1239
- fixes #1470
- fixes #1030
- fixes #951

### Checklist

- [X] All tests were successful.
- [X] Unit tests added.
- [] Documentation updated.

### CODE REVIEW

**Review by wajihyassine — COMMENTED** [review]
LGTM from web ui side per our side chats around it. As discussed, will leave the truncate of statuses as is to fix issues with table being too long and look at a later time for a way to size the virtual scroll to the table width.

**Comment by aarontp** [comment]
Quick update: I started reading through this today and will work on finishing the review tomorrow!

File: turbinia/state_manager.py [hunk 1]
```diff
@@ -547,3 +527,113 @@ def get_evidence_by_hash(self, file_hash: str) -> dict:
...
+        request_dict['last_update'] = json.dumps(
+            datetime.now().strftime(DATETIME_FORMAT))
+      #request_dict['status'] = json.dumps(
+      #    f'Task scheduled at {datetime.now().strftime(DATETIME_FORMAT)}')
+    except redis.RedisError as exception:
+      error_message = f'Error encoding key {request_key} in Redis'
+      log.error(f'{error_message}: {exception}')
+      raise TurbiniaException(error_message) from exception
+    # Either updates or write new key
+    if update == self.redis_client.key_exists(request_key):
```
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/state_manager.py:567]: Looks like there is similar logic in `write_evidence()`, so if anything changes here we might want to update similarly there.

**Review by aarontp — COMMENTED** [review]
Overall LGTM, just some small nits/comments.

File: turbinia/state_manager.py [hunk 2]
```diff
@@ -547,3 +527,113 @@ def get_evidence_by_hash(self, file_hash: str) -> dict:
...
+    try:
+      for (
+          attribute_name,
+          attribute_value) in self.redis_client.iterate_attributes(request_key):
+        request_dict[attribute_name] = attribute_value
+      request_dict['last_update'] = datetime.strptime(
+          request_dict.get('last_update'), DATETIME_FORMAT)
+      request_dict['start_time'] = datetime.strptime(
+          request_dict.get('start_time'), DATETIME_FORMAT)
+    except TurbiniaException as exception:
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/state_manager.py:591]: We are not raising an exception here so I don't think it's necessary.

File: turbinia/state_manager.py [hunk 3]
```diff
@@ -547,3 +527,113 @@ def get_evidence_by_hash(self, file_hash: str) -> dict:
...
+    try:
+      if not request_dict.get('last_update'):
+        request_dict['start_time'] = json.dumps(
+            datetime.now().strftime(DATETIME_FORMAT))
+      if not request_dict.get('last_update'):
+        request_dict['last_update'] = json.dumps(
+            datetime.now().strftime(DATETIME_FORMAT))
+      #request_dict['status'] = json.dumps(
+      #    f'Task scheduled at {datetime.now().strftime(DATETIME_FORMAT)}')
+    except redis.RedisError as exception:
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/state_manager.py:583]: Refactored this a bit.

File: turbinia/state_manager.py [hunk 4]
```diff
@@ -191,230 +222,167 @@ def get_task_data(
...
+  def write_new_task(self, task) -> Optional[str]:
+    """Writes task into redis.
 
     Args:
-      redis_key (str): The key of the Turbinia hash object in redis.
-      attribute_name (str): The name of the attribute to be get.
-      decode_json (bool): Boolean specifying if the value should be loaded.
+      task_dict (dict[str]): A dictionary containing the serialized
+        request attributes that will be saved.
+      update (bool): Allows overwriting previous key and blocks writing new
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/state_manager.py:312]: Done

File: turbinia/state_manager.py [hunk 5]
```diff
@@ -191,230 +222,167 @@ def get_task_data(
...
+      self.update_request_task(task)
+      # Set the current status for the TurbiniaRequest
+      request_status = self.get_request_status(task.request_id)
+      self.redis_client.set_attribute(
+          f'TurbiniaRequest:{task.request_id}', 'status',
+          json.dumps(request_status))
+      log.debug(f'Updating metadata for task {task.name} with key {task.id}')
+      task_dict = self.update_task_helper(task)
+      self.redis_client.write_hash_object(task_key, task_dict)
+    except TurbiniaException as exception:
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/state_manager.py:386]: Done

File: turbinia/state_manager.py [hunk 6]
```diff
@@ -547,3 +527,113 @@ def get_evidence_by_hash(self, file_hash: str) -> dict:
     """
     evidence_id = self.get_evidence_key_by_hash(file_hash).split(':')[1]
     return self.get_evidence_data(evidence_id)
+
+  def write_request(self, request_dict: dict, update=False):
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/state_manager.py]: I added a note to the ```TurbiniaRequest``` and ```Evidence``` docstrings to reflect how the objects are persisted to the datastore.

File: turbinia/api/models/request_status.py [hunk 7]
```diff
@@ -54,111 +53,29 @@ def get_request_data(
...
-      #    Turbinia server but not picked up by a worker yet.)
-      #
-      # Note that this method is concerned with a Turbiania request's status
-      # which is different than the status of an individual task.
-      self.status = 'running'
+      self.tasks = tasks if tasks else state_client.get_task_data(
+          instance=turbinia_config.INSTANCE_ID, request_id=request_id)
+
+    # Gets the information from the request if it is stored in Redis
+    if state_client.redis_client.key_exists(f'TurbiniaRequest:{request_id}'):
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/api/models/request_status.py:64]: Done. The `RedisClient` class handles the key name generation now.

File: turbinia/state_manager.py [hunk 8]
```diff
@@ -547,3 +527,113 @@ def get_evidence_by_hash(self, file_hash: str) -> dict:
...
+      output (str): Output of the function (keys | content | count).
+
+    Returns:
+      query_result (list | int): Result of the query. 
+    """
+    keys = []
+    for request_key in self.redis_client.iterate_keys('request'):
+      if stored_value := self.redis_client.get_attribute(request_key,
+                                                         attribute_name):
+        if (attribute_name == 'evidence_ids' and attribute_value in stored_value
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/state_manager.py]: Done

File: turbinia/workers/dfdewey.py [hunk 9]
```diff
@@ -93,7 +91,7 @@ def run(self, evidence, result):
         result.log(status_summary)
     else:
       status_summary = (
-          'Not running dfDewey. Case was not provided in task config.')
+          'Not running dfDewey. Case was not provided in taskconfig.')
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/workers/dfdewey.py]: Done

File: turbinia/state_manager.py [hunk 10]
```diff
@@ -547,3 +527,113 @@ def get_evidence_by_hash(self, file_hash: str) -> dict:
...
+    """
+    try:
+      request_key = ':'.join(
+          ('TurbiniaRequest', json.loads(request_dict['request_id'])))
+    except (TypeError, ValueError) as exception:
+      error_message = 'Error deserializing request attribute.'
+      log.error(f'{error_message}: {exception}')
+      raise TurbiniaException(error_message) from exception
+    try:
+      if not request_dict.get('last_update'):
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/state_manager.py]: Good catch, fixed.

File: turbinia/state_manager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/state_manager.py]: I refactored this a bit

File: turbinia/workers/__init__.py [hunk 11]
```diff
@@ -356,7 +336,7 @@ def add_evidence(self, evidence, evidence_config):
       return
 
     # We want to enforce this here to make sure that any new Evidence objects
-    # created also contain the config.  We could create a closure to do this
+    # created also contain theconfig.  We could create a closure to do this
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/workers/__init__.py]: Done

File: turbinia/workers/__init__.py [hunk 12]
```diff
@@ -253,7 +250,7 @@ def close(self, task, success, status=None):
             self.input_evidence.name, exception)
         self.log(
             message, level=logging.ERROR, traceback_=traceback.format_exc())
-        with filelock.FileLock(config.RESOURCE_FILE_LOCK):
+        with filelock.FileLock(config.CONFIG.RESOURCE_FILE_LOCK):
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/workers/__init__.py]: That was an unintended change, i reverted it.

File: turbinia/turbiniactl.py [hunk 13]
```diff
@@ -137,7 +137,7 @@ def process_args(args):
     else:
       config.LoadConfig()
   except TurbiniaException as exception:
-    print(f'Could not load config file ({exception!s}).\n{config.CONFIG_MSG:s}')
+    print(f'Could not load config file ({exception!s}).\n{config._MSG:s}')
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/turbiniactl.py]: Done

File: turbinia/task_manager.py [hunk 14]
```diff
@@ -545,7 +545,7 @@ def process_job(self, job, task):
       task (TurbiniaTask): The Task that just completed.
     """
     log.debug(f'Processing Job {job.name:s} for completed Task {task.id:s}')
-    self.state_manager.update_task(task)
+    # self.state_manager.update_task(task)
```
[the remark this answers is not in the document]
Comment (jleaniz) [reply turbinia/task_manager.py]: Done

File: turbinia/state_manager.py [hunk 15]
```diff
@@ -399,32 +408,42 @@ def write_evidence(self, evidence_dict: dict[str], update=False) -> str:
...
+    # and it can be quite verbose if a complex recipe is used.
+    try:
+      evidence_dict.pop('config')
+    except KeyError:
+      # Nothing to do if the key doesn't exist.
+      pass
+
+    try:
+      # Either updates or writes new key
+      if update == self.redis_client.key_exists(evidence_key):
```
Comment (aarontp) [inline turbinia/state_manager.py:434]: I'm still not sure I get this logic.  If update==True and the key does not exist, then we still won't write the key.   Maybe I'm missing something?

**Review by aarontp — APPROVED** [review]
LGTM, just a couple nits in comments and a question.

**Pull request merged**